README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# FlexVarJM

<!-- badges: start -->

[![R-CMD-check](https://github.com/LeonieCourcoul/FlexVarJM/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/LeonieCourcoul/FlexVarJM/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of FlexVarJM is to estimate joint model with subject-specific
time-dependent variability.

The global function is ‘lsjm’. It handles to estimate joint model with a
marker which has a subject-specific time-dependent variability and
competing events with the possibility to take into account the left
truncation. For more information you can read the corresponding article
: <https://arxiv.org/abs/2306.16785>

## Installation

You can install the development version of FlexVarJM from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("LeonieCourcoul/FlexVarJM")
```

## Example

# Estimation

This is an example in a simulated dataset where is a binary variable.

$$y_i(t_{ij}) = \color{blue}\tilde{y}_i(t_{ij}) \color{black} + \epsilon_{ij} = \beta_0 + b_{0i} + (\beta_1 + b_{1i})t_{ij} + \beta_2 * binary_i + \epsilon_{ij} $$

For the first risk, k = 1, we estimate the following risk function :

$$ \lambda_{i1}(t) = \lambda_{01}(t)\exp(\gamma_{11}*binary_i + \color{blue}\alpha_{11}\tilde{y}_i(t) + \color{red}\alpha_{\sigma 1} \sigma_i(t) \color{black}) $$
And for the second risk, k = 2 :
$$ \lambda_{i2}(t) = \lambda_{02}(t)\exp(\color{blue}\alpha_{21}\tilde{y_i}(t) + \color{blue}\alpha_{22}\tilde{y}'_i(t) + \color{red}\alpha_{\sigma 2} \sigma_i(t) \color{black}) $$

where :

- $\epsilon_{i}(t_{ij}) \sim \mathcal{N}(0, \color{red}\sigma_i^2\color{black})$
  with
  $\color{red}\log(\sigma_i(t_{ij})) = \mu_0 + \tau_{0i} + (\mu_1 + \tau_{1i})\times t_{ij} + \mu_2 * binary_i$

- with $b_i=\left(b_{0i},b_{1i}\right)^{\top}$ and
  $\tau_i=\left(\tau_{0i},\tau_{1i}\right)^{\top}$ assuming that the two
  sets of random effects $b_i$ and $\tau_i$ are not independent:
  $$(b_i, \tau_i)^\top \sim N(0, \Sigma)$$

- $\lambda_{0k}(t) = \kappa_k^2 t^{\kappa_k^2-1}e^{\zeta_{0k}}$ :
  Weibull function

- $\tilde{y}'_i(t)$ is the current slope of the marker $y$

``` r
example <- lsjm(formFixed = y~visit+binary,
                      formRandom = ~ visit,
                      formGroup = ~ID,
                      formSurv = Surv(time, event ==1 ) ~ binary,
                      timeVar = "visit",
                      data.long = Data_toy,
                      variability_hetero = TRUE,
                      formFixedVar =~visit+binary,
                      formRandomVar =~visit,
                      correlated_re = TRUE,
                      sharedtype = c("current value", "variability"),
                      hazard_baseline = "Weibull",
                      competing_risk = TRUE,
                      formSurv_CR = Surv(time, event ==2 ) ~ 1,
                      hazard_baseline_CR = "Weibull",
                      sharedtype_CR = c("slope", "variability"),
                      formSlopeFixed =~1,
                      formSlopeRandom = ~1,
                      indices_beta_slope = c(2), 
                      S1 = 500,
                      S2 = 1000,
                      nproc = 5,
                      Comp.Rcpp = TRUE
                      )
                      
summary(example)
```

You can access to the table of estimations and standard deviation with :

``` r
example$table.res
```

The computing time is given by :

``` r
example$time.compute
```

The output of the marqLevAlg algorithm is in :

``` r
example$result
```

Finally, some elements of control are in :

``` r
example$control
```

# Goodness-of-fit

You can check the goodness-of-fit of the longitudinal submodel and of
the survival submodel by computing the predicted random effects :

``` r
goodness <- goodness_of_fit(example, graph = T)
```

# Predictions

You can compute the probability for (new) individual(s) to have event 1
or 2 between time s and time s+t years given that he did not experience
any event before time s, its trajectory of marker until time s ans the
set of estimated parameters. To have a ‘IC%’ confidence interval, the
predictions are computed ‘nb.draws’ time and the percentiles of the
predictions are computed. For example, for individuals 1 and 3 to
experiment the event 1 at time 1.5, 2, and 3, given their measurements
until time 1 :

``` r
newdata <- Data_toy[which(Data_toy$ID %in% c(1,3)),]
predyn(newdata,example,1, c(1.5,2,3), event = 1, IC = 95, nb.draws = 500, graph = TRUE)
```


NEWS.md
# FlexVarJM 0.1.0

* Initial CRAN submission.


R/predyn.R
#' Dynamic prediction for a new individual
#'
#' @param newdata data frame : collected data for a new individual
#' @param object lsjm object : estimation of the model
#' @param s numeric : the time to begin prediction
#' @param times numeric vector : future times to calculate predictions
#' @param event integer (0, 1 or 2) : the event of interest for the prediction
#' @param IC integer : percentage of confidence for the interval confidence (between 0 and 100), 95 by default, NULL if no IC
#' @param nb.draws integer : the number of simulations to compute the interval confidence (by bootstrap), 500 by default
#' @param graph boolean : indicator to plot the graphs or not
#'
#' @return A table of dynamic predictions
#'
#' @examples
#' \donttest{
#' 
#'
#' #fit a joint model with competing risks and subject-specific variability
#' example <- lsjm(formFixed = y~visit,
#' formRandom = ~ visit,
#' formGroup = ~ID,
#' formSurv = Surv(time, event ==1 ) ~ 1,
#' timeVar = "visit",
#' data.long = Data_toy,
#' variability_hetero = TRUE,
#' formFixedVar =~visit,
#' formRandomVar =~visit,
#' correlated_re = TRUE,
#' sharedtype = c("current value", "variability"),
#' hazard_baseline = "Weibull",
#' formSlopeFixed =~1,
#' formSlopeRandom = ~1,
#' indices_beta_slope = c(2), 
#' competing_risk = TRUE,
#' formSurv_CR = Surv(time, event ==2 ) ~ 1,
#' hazard_baseline_CR = "Weibull",
#' sharedtype_CR = c("current value", "variability"),
#' S1 = 100,
#' S2 = 1000,
#' nproc = 1,
#' maxiter = 100,
#' Comp.Rcpp = TRUE
#' )
#' 
#' #Prediction for individuals 1 and 3 to experiment the event 1 
#' #at time 1.5, 2, and 3, given their measurements until time 1:
#' newdata <- Data_toy[which(Data_toy$ID %in% c(1,3)),]
#' pred.new <- predyn(newdata,example,1, c(1.5,2,2.8,3), event = 1, IC = 95, 
#' nb.draws = 100, graph = TRUE)
#' }
#' @export

predyn <- function(newdata, object, s, times, event = 1, IC = 95, nb.draws = 500, graph = FALSE){
  if(!inherits(object, "lsjm")) stop("use only \"lsjm\" objects")
  if(object$result$istop != 1) stop("The estimation didn't reach convergence \n")
  if(IC<=0 || IC>=100) stop("IC must be between 0 and 100")
  if(!is.null(IC) && (is.null(nb.draws) || nb.draws <=0)) stop("draw must be higher 1")
  bootstrap <- c()
  pred.ponct <- c()
  id <- as.integer(newdata[all.vars(object$control$formGroup)][,1])
  newdata$id <- id
  window <- times - s
  table.pred <- c()
  if(!is.null(IC)){
    for(i in unique(id)){
      newdata.id <- newdata[which(newdata$id == i),]
      bootstrap <- c()
      pred.ponct <- c()
      for(t in window){
        pred.t <- pred_s.t.ponctuel.tps(newdata.id,object,s,t,event)
        pred.t.bootstrap.tps <- pred_s.t.bootstrap.tps(newdata.id,object,s,t,event,nb.draws )
        pred.ponct <- c(pred.ponct, pred.t)
        bootstrap<- cbind(bootstrap,pred.t.bootstrap.tps)
      }
      table.pred.id <- cbind(i, times, pred.ponct, apply(bootstrap,2, function(x) quantile(x, 0.50)),
                             apply(bootstrap,2, function(x) quantile(x, ((100-IC)/2)/100)),
                             apply(bootstrap,2, function(x) quantile(x, 1-((100-IC)/2)/100)),
                             apply(bootstrap,2, sd)
      )
      table.pred.id <- as.data.frame(table.pred.id)
      colnames(table.pred.id) <- c("ID","Time","Prediction","Median","ICinf","ICsup", "Empirical SD")
      if(graph){
        #browser()
        data.long.until.time.s <-subset(newdata.id, get(object$control$timeVar)<=s)
        x.axe <- c(data.long.until.time.s[,all.vars(object$control$formFixed)[2]],times)
        #print(x.axe)
        y.axe <- c(data.long.until.time.s[,all.vars(object$control$formFixed)[1]], rep(NA,length(window)))
        #print(y.axe)
        y.axe2 <- c(rep(NA,length(data.long.until.time.s[,all.vars(object$control$formFixed)[2]])),table.pred.id$ICinf)
        #print(y.axe2)
        y.axe3 <- c(rep(NA,length(data.long.until.time.s[,all.vars(object$control$formFixed)[2]])),table.pred.id$Median)
        #print(y.axe3)
        y.axe4 <- c(rep(NA,length(data.long.until.time.s[,all.vars(object$control$formFixed)[2]])),table.pred.id$ICsup)
        y.axe5 <- c(rep(NA,length(data.long.until.time.s[,all.vars(object$control$formFixed)[2]])),table.pred.id$Prediction)
        plot(x = x.axe, y = y.axe,xlim = c(0,max(s+window)),
             xlab = "Time", ylab = "Marker", cex.lab = 1, col = "black",
             main = "Prediction of event", pch = 20, cex = 1, font = 1, font.lab = 1, cex.lab = 1, cex.main = 1)
        graphics::par(new = TRUE, font = 1, cex.lab = 1)
        plot(x.axe, y.axe3, axes = FALSE,col = "black", type = "l", ylim = c(0.000001, max(table.pred.id$ICsup, na.rm = T)),ylab = "", xlab = "", lwd =2, font.lab = 1, cex.lab = 1 )
        graphics::lines(x.axe, y.axe5, col = "red", lty=1, lwd = 2)
        graphics::lines(x.axe, y.axe2, col = "black", lty=2, lwd = 2)
        graphics::lines(x.axe, y.axe4, col = "black", lty=2, lwd = 2)
        graphics::axis(side= 4, cex = 2)
        graphics::abline(v = s, lty = 3)
        graphics::mtext("Probability of event", cex = 1, side = 4, line = 3, font.lab = 1)
        
      }
      table.pred <- rbind(table.pred, table.pred.id)
    }
    table.pred <- as.data.frame(table.pred)
    colnames(table.pred) <- c("ID","Time","Prediction","Median","ICinf","ICsup", "Empirical SD")
    graph.predyn <- NULL
  }
  else{
    for(i in unique(id)){
      newdata.id <- newdata[which(newdata$id == i),]
      bootstrap <- c()
      pred.ponct <- c()
      for(t in window){
        pred.t <- pred_s.t.ponctuel.tps(newdata.id,object,s,t,event)
        pred.ponct <- c(pred.ponct, pred.t)
      }
      table.pred.id <- cbind(i, times, pred.ponct
      )
      table.pred <- rbind(table.pred, table.pred.id)
    }
    table.pred <- as.data.frame(table.pred)
    colnames(table.pred) <- c("ID","Time","Prediction")
    graph.predyn <- NULL
    if(graph){
      data.long.until.time.s <-subset(newdata.id, get(object$control$timeVar)<=s)
      x.axe <- c(data.long.until.time.s[,all.vars(object$control$formFixed)[2]],times)
      #print(x.axe)
      y.axe <- c(data.long.until.time.s[,all.vars(object$control$formFixed)[1]], rep(NA,length(window)))
      y.axe5 <- c(rep(NA,length(data.long.until.time.s[,all.vars(object$control$formFixed)[2]])),table.pred.id$Prediction)
      plot(x = x.axe, y = y.axe,xlim = c(0,s+window),
           xlab = "Time", ylab = "Marker", cex.lab = 1, col = "black",
           main = "Prediction of event", pch = 20, cex = 1, font = 1, font.lab = 1, cex.lab = 1, cex.main = 1)
      graphics::par(new = TRUE, font = 1, cex.lab = 1)
      plot(x.axe, y.axe5, axes = FALSE,col = "black", type = "l", ylim = c(0.000001, max(table.pred.id$ICsup, na.rm = T)),ylab = "", xlab = "", lwd =2, font.lab = 1, cex.lab = 1 )
      #graphics::lines(x.axe, y.axe5, col = "red", lty=2, lwd = 2)
      #graphics::lines(x.axe, y.axe2, col = "black", lty=2, lwd = 2)
      #graphics::lines(x.axe, y.axe4, col = "black", lty=2, lwd = 2)
      graphics::axis(side= 4, cex = 2)
      graphics::abline(v = s, lty = 3)
      graphics::mtext("Probability of event", cex = 1, side = 4, line = 3, font.lab = 1)
      
    }
  }
  

  result <- table.pred
  result
}
